File: 11/stones.py
from collections import defaultdict
#
# IN = "5910927 0 1 47 261223 94788 545 7771"
IN = "125 17"

stones = [int(x) for x in IN.split()]

cache = defaultdict(defaultdict)

for _ in range(25):
    i = 0
    while i < len(stones):
        stone = stones[i]
        if stone == 0:
            stones[i] = 1
            i += 1
        elif len(str(stone)) % 2 == 0:
            del stones[i]
            s = str(stone)
            mid = int(len(s)/2)
            left, right = s[:mid], s[mid:]
            stones.insert(i, int(right))
            stones.insert(i, int(left))
            i += 2
        else:
            stones[i] = stone * 2024
            i += 1

print('stones', len(stones))
assert len(stones) == 229557103025807

File: 11/stones.go
package main2

import (
	"fmt"
	"strconv"
)

func blink(stones []int, blinks int) <-chan int {
	// fmt.Printf("%d %+v\n", blinks, stones)
	ch := make(chan int)

	go func() {
		defer close(ch)

		if blinks == 0 {
			for _, stone := range stones {
				ch <- stone
			}
			return
		}

		for _, stone := range(stones) {
			if stone == 0 {
				// fmt.Printf("%d, ->1\n", stone)
				for s := range blink([]int{1}, blinks - 1) {
					ch <- s
				}
			} else if s := strconv.Itoa(stone); len(s) % 2 == 0 {
				// fmt.Printf("%d, split\n", stone)
				mid := int(len(s)/2)
				left, _ := strconv.Atoi(s[:mid])
				right, _ := strconv.Atoi(s[mid:])
				for s := range blink([]int{left, right}, blinks - 1) {
					ch <- s
				}
			} else {
				// fmt.Printf("%d, *2024\n", stone)
				for stone := range blink([]int{stone*2024}, blinks -1) {
					ch <- stone
				}
			}
		}

		// for _, stone := range stones {
		// 	ch <- stone
		// }

		// stones
	}()
	return ch
}

func main() {
	var stones []int

	// stones = []int{125,17}
	stones = []int{5910927,0,1,47,261223,94788,545,7771}

	blinks := blink(stones, 75)

	count := 0
	for _ = range blinks {
		// fmt.Printf("%d ", stone)
		count++
	}
	fmt.Println()
	// fmt.Printf("stones %s\n", stones)
	fmt.Printf("stones %d\n", count)
}
